### PR TITLE
feat(cli): support migrating state between backends

### LINKED ISSUE #489: Handle Switching Backend from Remote to Something Else
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When the Remote Backend is dropped in favour of something else (e.g. Local), the Terraform CLI detects this change and asks if the existing state should be copied, see:

Screenshot 2021-01-13 at 14 53 04

The `init` command has a `force-copy` flag, which would essentially confirm this question with a yes. What do you think, is it safe to assume it's ok confirming this all the time and make it a default flag on the `init` step? 

### References

- #449

### LINKED ISSUE #490: Auto Migrate State into Terraform Cloud
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We're aiming to improve the usability for Terraform Cloud scenarios in cdktf. One part of this is the state migration into the Terraform Cloud. There are two scenarios:

1. Fresh / empty remote workspace
2. Remote workspace with pre-existing state

While it seems safe to assume that it's ok to auto-migrate state in the first case, we need to confirm with the user in the second case. 

The Terraform CLI is confirming even in he first case, and erroring out in the 2nd case.

We'll need to think about the full process and potential edge cases, if we really want to do this automatically - even when confirming with the user.

##### No State

Screenshot 2021-01-13 at 15 34 29

#####  Existing State

Screenshot 2021-01-13 at 15 25 41

### References

- #449

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
After this change a --migrate-state flag can be passed to diff, deploy, destroy
that will migrate the state from the previous backend to the current one.

Please not that Terraform CLI does not support migrating from the cloud configuration
to other backends, as a result this is not supported in cdktf either.

Closes #490
Closes #489

### CODE REVIEW

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts [hunk 1]
```diff
@@ -106,28 +107,51 @@ export class TerraformCli implements Terraform {
         createTerraformLogHandler(phase, filter)(stderr.toString(), true);
   }
 
-  public async init(needsUpgrade: boolean, noColor?: boolean): Promise<void> {
+  public async init(
+    needsUpgrade: boolean,
+    noColor: boolean,
+    migrateState: boolean
```
Comment (ansgarm) [inline packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:113]: nit: We're slowly approaching a count of args here, where I'd prefer a config object over a list of args 😅

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts [hunk 2]
```diff
@@ -106,28 +107,51 @@ export class TerraformCli implements Terraform {
...
+      (data) => {
+        stdout(data);
+        if (data.includes("Should Terraform migrate your existing state?")) {
+          if (migrateState) {
+            actions.writeLine("yes");
+          } else {
+            actions.writeLine("no");
+            stderr(
+              "Please pass the --migrate-state flag to migrate your state"
+            );
```
Comment (ansgarm) [inline packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:145]: Nice!

Will `terraform init` return with a non-zero exit code if we send "no"? If I remember correctly `terraform apply` does so when we reject the plan.

File: packages/cdktf-cli/src/bin/cmds/deploy.ts [hunk 3]
```diff
@@ -84,6 +84,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/deploy.ts:91]: Maybe reword this to
```suggestion
        desc: "Pass this flag after switching state backends to approve a state migration for all targeted stacks",
```

File: packages/cdktf-cli/src/bin/cmds/destroy.ts [hunk 4]
```diff
@@ -73,6 +73,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/destroy.ts:80]: Maybe reword as proposed above?
```suggestion
        desc: "Pass this flag after switching state backends to approve a state migration for all targeted stacks",
```

File: packages/cdktf-cli/src/bin/cmds/diff.ts [hunk 5]
```diff
@@ -65,6 +65,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for the selected stack.",
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/diff.ts:72]: Reeeeword, too?
```suggestion
        desc: "Pass this flag after switching state backends to approve a state migration for the targeted stack",
```

**Review by ansgarm — APPROVED** [review]
Looking good!
I have a proposed rewording for the flag description, but no strong feelings there

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts (hunk 2 above)
Comment (DanielMSchmidt) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:145]: Hmm, I'm unsure. I think I'll just cancel the init and throw an error, thats safer
